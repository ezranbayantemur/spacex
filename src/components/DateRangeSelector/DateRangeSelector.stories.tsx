import {View} from 'react-native';
import {color} from '@style';
import DateRangeSelector from './DateRangeSelector';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {DateRangeSelectorProps} from './DateRangeSelector.types';

const DateRangeSelectorMeta: Meta<DateRangeSelectorProps> = {
  argTypes: {
    onSearch: {
      action: 'on selection',
    },
  },
  title: 'DateRangeSelector',
  component: DateRangeSelector,
  decorators: [
    Story => (
      <View style={{flex: 1, backgroundColor: color.BLACK}}>
        <Story />
      </View>
    ),
  ],
};

export default DateRangeSelectorMeta;

export const Basic: StoryObj<DateRangeSelectorProps> = {};
