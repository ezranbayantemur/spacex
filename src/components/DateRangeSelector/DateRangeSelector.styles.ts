import {StyleSheet} from 'react-native';
import {color, font, padding, typography} from '@style';

export default StyleSheet.create({
  container: {
    marginTop: padding.MEDIUM,
    padding: padding.MEDIUM,
    paddingHorizontal: padding.LARGE,
  },
  title: {
    fontSize: typography.NORMAL,
    fontFamily: font.REGULAR,
    color: color.ALABASTER,
    alignSelf: 'center',
  },
  spacer: {
    marginTop: padding.MEDIUM,
  },
});
