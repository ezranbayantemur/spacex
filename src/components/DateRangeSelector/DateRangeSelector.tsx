import React from 'react';
import {Text, View} from 'react-native';
import {isEqual} from 'date-fns';
import Button from '../Button';
import DatePicker from '../DatePicker';
import styles from './DateRangeSelector.styles';
import type {DateRangeSelectorProps} from './DateRangeSelector.types';

const DateRangeSelector = ({testID, onSearch}: DateRangeSelectorProps) => {
  const [startDate, setStartDate] = React.useState<Date>(new Date());
  const [endDate, setEndDate] = React.useState<Date>(new Date());

  const handleOnSearch = () => {
    if (isEqual(startDate, endDate)) {
      return;
    }

    onSearch({
      start: startDate,
      end: endDate,
    });
  };

  return (
    <View style={styles.container}>
      <DatePicker
        testID={`${testID}_start`}
        maximumDate={endDate}
        onDateSelected={setStartDate}
      />
      <Text style={styles.title}>to</Text>
      <DatePicker
        testID={`${testID}_end`}
        minimumDate={startDate}
        onDateSelected={setEndDate}
      />
      <View style={styles.spacer} />
      <Button
        testID={`${testID}`}
        text="Search"
        onPress={handleOnSearch}
      />
    </View>
  );
};

export default DateRangeSelector;
