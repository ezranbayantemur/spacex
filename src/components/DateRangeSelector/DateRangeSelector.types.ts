export interface OnSearchCallbackProps {
  start: Date;
  end: Date;
}

export interface DateRangeSelectorProps {
  onSearch: (dates: OnSearchCallbackProps) => void;
}
