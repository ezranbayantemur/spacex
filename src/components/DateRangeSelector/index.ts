export * from './DateRangeSelector.types';

export {default} from './DateRangeSelector';
