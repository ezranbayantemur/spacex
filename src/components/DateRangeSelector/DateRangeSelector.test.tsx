import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import DateRangeSelector from './DateRangeSelector';

jest.useFakeTimers().setSystemTime(new Date('2010-01-01'));
const mockOnDateSelected = jest.fn();
jest.mock('react-native-date-picker', () => 'RNDatePicker');

let wrapper: ReturnType<typeof render>;

describe('DateRangeSelector unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <DateRangeSelector
        testID="daterangepicker"
        onSearch={mockOnDateSelected}
      />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should return correctly date range', () => {
    const start = wrapper.getByTestId('daterangepicker_start_picker');
    fireEvent(start, 'onDateSelected', new Date('2020-01-01'));

    const end = wrapper.getByTestId('daterangepicker_end_picker');
    fireEvent(end, 'onDateSelected', new Date('2021-01-01'));

    const button = wrapper.getByTestId('daterangepicker_button_touchable');
    fireEvent(button, 'onPress');

    expect(mockOnDateSelected).toHaveBeenCalledWith({
      start: new Date('2020-01-01'),
      end: new Date('2021-01-01'),
    });
  });
});
