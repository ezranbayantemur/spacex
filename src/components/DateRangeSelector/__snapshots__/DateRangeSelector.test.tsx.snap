// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangeSelector unit tests should render correctly 1`] = `
<View
  style={
    {
      "marginTop": 16,
      "padding": 16,
      "paddingHorizontal": 24,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "#FBFBFB",
        "borderWidth": 2,
        "justifyContent": "center",
        "margin": 8,
        "opacity": 1,
        "padding": 8,
      }
    }
    testID="daterangepicker_start_touchable"
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Quicksand-Bold",
          "fontSize": 17,
        }
      }
    >
      01 Jan 2010
    </Text>
    <RNDatePicker
      date={2010-01-01T00:00:00.000Z}
      maximumDate={2010-01-01T00:00:00.000Z}
      modal={true}
      mode="date"
      onCancel={[Function]}
      onConfirm={[Function]}
      open={false}
      testID="daterangepicker_start_picker"
    />
  </View>
  <Text
    style={
      {
        "alignSelf": "center",
        "color": "#FBFBFB",
        "fontFamily": "Quicksand-Regular",
        "fontSize": 15,
      }
    }
  >
    to
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "#FBFBFB",
        "borderWidth": 2,
        "justifyContent": "center",
        "margin": 8,
        "opacity": 1,
        "padding": 8,
      }
    }
    testID="daterangepicker_end_touchable"
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Quicksand-Bold",
          "fontSize": 17,
        }
      }
    >
      01 Jan 2010
    </Text>
    <RNDatePicker
      date={2010-01-01T00:00:00.000Z}
      minimumDate={2010-01-01T00:00:00.000Z}
      modal={true}
      mode="date"
      onCancel={[Function]}
      onConfirm={[Function]}
      open={false}
      testID="daterangepicker_end_picker"
    />
  </View>
  <View
    style={
      {
        "marginTop": 16,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#176B87",
        "margin": 8,
        "opacity": 1,
        "padding": 16,
        "paddingHorizontal": 32,
      }
    }
    testID="daterangepicker_button_touchable"
  >
    <Text
      style={
        {
          "color": "#FBFBFB",
          "fontFamily": "Quicksand-Bold",
          "fontSize": 15,
        }
      }
      testID="daterangepicker_button_text"
    >
      Search
    </Text>
  </View>
</View>
`;
